### PR TITLE
Refactor session tests

### PR DESCRIPTION
**Pull request by sietseringers**
This solves the following issues for the session tests in `internal/sessiontest`:

1. Removal of all global state by the session tests, resulting in code that is simpler and better to maintain.
2. Unneccesary distinctions and near-code duplication has been removed. There used to be not one but two helper functions that performed test sessions, one of them using the IRMA server and another the IRMA server library. Some of the tests used the one while others used the other, often without good reason: a simple disclosure session should work against both the IRMA server and library. Now there is a single helper function which can work against both the server and library, and many session tests are now in fact run against both.

The second point is achieved as follows. Taking the simplest session test as example, the function
```go
func TestDisclosureSession(t *testing.T)
```

has changed to the following:
```go
func testDisclosureSession(t *testing.T, conf interface{}, opts ...sessionOption)
```

Here, the `conf` parameter must be either of type `func() *requestorserver.Configuration` or `func() *server.Configuration`. The type determines if an IRMA server or IRMA server library is used, respectively. During the test, the function is invoked to obtain the configuration that is used within the test.

The `testDisclosureSession` cannot be started as a Go unit test directly because it does not have the correct signature, but it can be converted into a function that does using the new `curry` function. For example, this runs this test using first an IRMA server and then again using an IRMA server library:

```go
t.Run("DisclosureSession", curry(testDisclosureSession, IrmaServerConfiguration))
t.Run("DisclosureSession", curry(testDisclosureSession, IrmaLibraryConfiguration))
```

The test may be further customized as follows: 
* Additional `sessionOption` constants may be passed as well to `curry`, which passes them along to `testDisclosureSession`, which passes them along to the session helper function.
* The server configuration functions (the second parameter to `curry`) may be customized using decorators, for example to enable Redis or to change scheme settings in the configuration.

### CODE REVIEW

File: internal/sessiontest/requestor_test.go [hunk 1]
```diff
@@ -42,67 +68,161 @@ func processOptions(options ...sessionOption) sessionOption {
...
+			return nil, nil, false
+		}
+		if conf != nil {
+			require.FailNow(t, "either specify irmaServer or conf")
+			return nil, nil, false
+		}
+		return irmaServer, nil, false
+	}
+	if opts.enabled(sessionOptionReuseServer) {
+		return nil, nil, false
```
Comment (synaptic-cleft) [inline internal/sessiontest/requestor_test.go:92]: Currently a configuration can be  passed to the `doSession` function in combination with `sessionOptionReuseServer`. However, the code will only work with the JWT configuration as no configuration is returned here.

This is fine as we currently don't have any test where we need to use a different configuration. So I suggest we error out here if a configuration is passed and add a comment where `sessionOptionReuseServer` is defined that this option can only be used with JWT configurations.

File: internal/sessiontest/main_test.go [hunk 2]
```diff
@@ -1,28 +1,23 @@
...
 	"github.com/privacybydesign/irmago/internal/test"
 	"github.com/privacybydesign/irmago/irmaclient"
-	"github.com/privacybydesign/irmago/server"
 	"github.com/stretchr/testify/require"
 )
 
 // Defines the maximum protocol version of an irmaclient in tests
-var maxClientVersion = &irma.ProtocolVersion{Major: 2, Minor: 8}
+// TODO: rename to prePairingClientVersion
+var maxClientVersion = &irma.ProtocolVersion{Major: 2, Minor: 7}
```
[the remark this answers is not in the document]
Comment (sietseringers) [reply internal/sessiontest/main_test.go]: Structs cannot be a const, so yes. Or did you mean to suggest to move its definition closer to its usage?

File: internal/sessiontest/requestor_test.go [hunk 3]
```diff
@@ -117,20 +304,20 @@ func waitSessionFinished(t *testing.T, irmaServer *irmaserver.Server, token irma
 }
 
 // Check that nonexistent IRMA identifiers in the session request fail the session
-func TestRequestorInvalidRequest(t *testing.T) {
-	StartIrmaServer(t, false, "")
-	defer StopIrmaServer()
-	_, _, _, err := irmaServer.StartSession(irma.NewDisclosureRequest(
+func TestInvalidRequest(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (sietseringers) [reply internal/sessiontest/requestor_test.go]: Looks like it, yes. At any rate, the `Requestor` in the name is of little use.

File: internal/sessiontest/session_test.go [hunk 4]
```diff
@@ -21,44 +22,91 @@ import (
...
+	t.Run("UnsatisfiableDisclosureSession", curry(testUnsatisfiableDisclosureSession, IrmaLibraryConfiguration))
+
+	t.Run("StaticQRSession", curry(testStaticQRSession, nil)) // has its own configuration
+}
+
+// curry takes (1) a test function which apart from *testing.T additionally accepting a
+// configuration function and session options, and (2) a configuration function and session objects,
+// and returns a function suitable for unit testing by applying the configuration function and
+// session options in the rightmost two parameter slots of the specified function.
+func curry(
```
[the remark this answers is not in the document]
Comment (sietseringers) [reply internal/sessiontest/session_test.go]: Renamed to `apply()`.

File: internal/sessiontest/main_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sietseringers) [reply internal/sessiontest/main_test.go]: Moved towards its only usage.

File: internal/sessiontest/requestor_test.go [hunk 5]
```diff
@@ -42,63 +68,224 @@ func processOptions(options ...sessionOption) sessionOption {
...
+			sesPkg  server.SessionPackage
+			err     error
+			useJWTs bool
+		)
+		if conf != nil {
+			useJWTs = !conf.(*requestorserver.Configuration).DisableRequestorAuthentication
+		} else {
+			useJWTs = true
+		}
+		url := "http://localhost:48682"
```
[the remark this answers is not in the document]
Comment (synaptic-cleft) [reply internal/sessiontest/requestor_test.go]: Would you mind still doing a global search on `4868` and set some other frequently used fixed ports next to this extracted one?

File: internal/sessiontest/server_test.go [hunk 6]
```diff
@@ -125,8 +122,15 @@ func chainedServerHandler(t *testing.T) http.Handler {
 		require.NoError(t, err)
 		require.NoError(t, r.Body.Close())
 
-		var result server.SessionResult
-		require.NoError(t, json.Unmarshal(bts, &result))
+		claims := &struct {
+			jwt.StandardClaims
```
[the remark this answers is not in the document]
Comment (synaptic-cleft) [reply internal/sessiontest/server_test.go]: In hindsight I saw that we use the `StandardClaims` in other places, too. In some places, it's not just test code. Should we for consistency also update them now?

File: internal/sessiontest/requestor_test.go (hunk 1 above)
Comment (synaptic-cleft) [reply internal/sessiontest/requestor_test.go]: Thanks!

I've just noticed that here and also above yo return after the require.FailNow. If I get FailNow correctly, the return statements will never be triggered and can be removed.

File: internal/sessiontest/server_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sietseringers) [reply internal/sessiontest/server_test.go]: We should update them, but I'm a little hesitant to do that in this PR, which so far eclusively deals with test code. I'd say we do that in a separate PR.

**Review by synaptic-cleft — APPROVED** [review]

**Pull request merged**